cpu/atxmega/periph/gpio.c

```diff
@@ -0,0 +1,510 @@
+/*
+ * Copyright (C) 2018 RWTH Aachen, Josua Arndt <jarndt@ias.rwth-aachen.de>
+ *               2021 Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_atxmega
+ * @ingroup     cpu_atxmega_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level GPIO driver implementation
+ *
+ * @author      Josua Arndt <jarndt@ias.rwth-aachen.de>
+ * @author      Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <avr/interrupt.h>
+
+#include "cpu.h"
+#include "periph_conf.h"
+#include "periph/gpio.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#define GPIO_PORT_BASE      ((uint16_t)&PORTA)
+#define GPIO_PORT_OFFSET    (0x20)
+
+static gpio_isr_ctx_t config_ctx[GPIO_EXT_INT_NUMOF];
+static uint8_t config_irq[GPIO_EXT_INT_NUMOF];
+
+/**
+ * @brief     Extract the pin number of the given pin
+ */
+static inline uint8_t _pin_mask(gpio_t pin)
+{
+    return (pin & 0xff);
+}
+
+/**
+ * @brief     Extract the port number of the given pin
+ */
+static inline uint16_t _port_num(gpio_t pin)
+{
+    return (pin >> 8) & 0x00ff;
+}
+
+/**
+ * @brief     Generate the PORTx address of the give pin.
+ */
+static inline uint16_t _port_addr(gpio_t pin)
+{
+    uint16_t port_num = _port_num(pin);
+    uint16_t port_addr = GPIO_PORT_BASE + (port_num * GPIO_PORT_OFFSET);
+
+    return port_addr;
+}
+
+#if ENABLE_DEBUG
+static inline void _print_config(gpio_t pin)
+{
+    PORT_t *port = (PORT_t *)_port_addr(pin);
+    uint8_t pin_mask = _pin_mask(pin);
+    volatile uint8_t *pin_ctrl = &port->PIN0CTRL;
+
+    DEBUG("PORT: 0x%04x, PIN: 0x%02x\n", (uint16_t)port, pin_mask);
+    DEBUG("DIR: 0x%02x, IN: 0x%02x, OUT: 0x%02x\n", port->DIR, port->IN, port->OUT);
+    DEBUG("INTCTRL: 0x%02x\nINTFLAGS: 0x%02x\n", port->INTCTRL, port->INTFLAGS);
+    DEBUG("INT0MASK: 0x%02x\nINT1MASK: 0x%02x\n", port->INT0MASK, port->INT1MASK);
+
+    for (uint8_t p = 0; p < 8; p++) {
+        DEBUG("PIN%dCTRL: 0x%02x\n", p, pin_ctrl[p]);
+    }
+}
+#endif
+
+static inline void _gpio_pinctrl_set(gpio_t pin, gpio_mode_t mode,
+                                     gpio_flank_t flank, gpio_cb_t cb,
+                                     void *arg)
+{
+    uint8_t pin_mask = _pin_mask(pin);
+    uint16_t port_num = _port_num(pin);
+    PORT_t *port = (PORT_t *)_port_addr(pin);
+    volatile uint8_t *pin_ctrl = &port->PIN0CTRL;
+    int irq_state;
+    uint8_t tmp;
+
+    if (mode & GPIO_OUT) {
+        port->DIRSET = pin_mask;
+    }
+    else {
+        port->DIRCLR = pin_mask;
+    }
+
+    irq_state = irq_disable();
+
+    for (uint8_t p = 0; p < 8; p++) {
+        if (pin_mask & (1 << p)) {
```

Comment:
I guess there is only ever one bit set in `pin_mask`, right?

```suggestion
    uint8_t p = bitarithm_lsb(pin_mask);
```

Reply:
I assume you want to allow multiple bits to be set in `pin_mask` as an architecture specific extension? Otherwise `bitarithm_lsb()` as suggested by @benpicco is easier to read and potentially faster.

Note that there is also [`bitarithm_test_and_clear()`](https://api.riot-os.org/bitarithm_8h.html#a90bfe935dcde09afd2918a3768ad2b17) that could be your friend here, if indeed more than one bit can be set.